### ISSUE TITLE
"Animator" objects are leaked if sprite is deleted before animation is complete

### ISSUE THREAD

**Issue by SirGnip** [maarten-dp/arcade-curtains#1]
Greetings!  I've been experimenting with curtain's animation.  Nice library! 

An observation.  I found that if an animation is playing and the sprite is destroyed before its animation completes, the associated `Animator` object continues to exist in the `AnimationManager.animations` list.  Not a huge deal for a simple example.  But if you have lots of sprites in your game with dynamic lifetimes, these unused animation objects hang around in the list and begin to degrade performance after awhile.

Here is a repro.  Run the app and look at the output spew.  While the animation is running, you can see it repeatedly prints `Length of animations: 1` as expected.  But, when you press the "X" key to delete the sprite, the output still says "1", meaning the `Animator` object was never removed from the list:

```
import time
import arcade
from arcade_curtains import Curtains, BaseScene

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600


class MyGame(arcade.Window):
    def __init__(self, width, height, title):
        super().__init__(width, height, title)
        arcade.set_background_color(arcade.color.BLACK)
        self.curtains = Curtains(self)
        self.curtains.add_scene('single', SingleScene(self))
        self.curtains.set_scene('single')

    def on_key_press(self, symbol: int, modifiers: int):
        if symbol == arcade.key.ESCAPE:
            self.close()
        if symbol == arcade.key.X:
            print('removed sprite')
            del self.curtains.current_scene.sprite
            self.curtains.current_scene.sprite = None


class SingleScene(BaseScene):
    def setup(self, win):
        self.sprite = arcade.Sprite(":resources:images/tiles/boxCrate_double.png")
        self.sprite.position = (100, 100)

    def enter_scene(self, previous_scene):
        if self.sprite:
            self.sprite.animate(
                duration=20,
                position=(500, 500),
            )

    def draw(self):
        time.sleep(0.1)
        print("Length of animations:", len(arcade.get_window().curtains.current_scene.animations.animations))
        if self.sprite:
            self.sprite.draw()


def main():
    window = MyGame(SCREEN_WIDTH, SCREEN_HEIGHT, "perf test")
    arcade.run()


if __name__ == "__main__":
    main()
```

Poking through the code, it seems this is because the only way an `Animator` object can be removed from that list is if `Animator.finished` is True.  In the case of deleting a sprite before its animation completes, this will never be true.  This "leak" also means that the sprite will never be garbage collected because the `Animator` has a reference to the sprite.  Thoughts on how to clean this up?  I don't see any way to navigate the data structures, starting with the sprite, and finding its specific `Animator` object in the `AnimationManager.animations` list, except for searching the entire list.

**Comment by maarten-dp** [comment]
Hi SirGnip,

I'm very pleased to see that people are having interest in this library, and are now making issues!
Thank you very much :)

I'll have a closer look at it this evening when I'll have more time.

**Comment by maarten-dp** [comment]
Hi SirGnip,

I had a closer look.

First I would like to discuss how you delete your sprite.
Using `del` to remove an object isn't an advised way to get rid of a sprite (or any object in python, for that matter). What you're doing is removing the reference to the object from your scene, and not really deleting the object itself.

Consider this example:
![image](https://user-images.githubusercontent.com/35668288/95224772-605faa00-07fb-11eb-95fe-a47e654696ff.png)
As you can see, I'm only deleting the reference that `a` is holding, I'm not actually deleting the list.
I believe the proper way to delete sprites is to call the [`kill` function](https://github.com/pythonarcade/arcade/blob/development/arcade/sprite.py#L988).

That being said, this still doesn't solve your problem (and it might even break the library if you do it like this :p)
So, my first impression is to overload the `kill` function, and remove any animations before executing the native kill in arcade.

Would that sound like an agreeable solution to you?

**Comment by SirGnip** [comment]
Sounds reasonable to me.  Adding some logic to `kill()` that cleans up no-longer-needed entries in `AnimationManager.animations` is a straightforward way to do it.

I was briefly considering solutions that used [weakref](https://docs.python.org/3/library/weakref.html), but that might be overkill.  Though, if I understand it properly, if AnimationManager used weakrefs to refer to sprites, client code wouldn't have to remember to call `kill()` on the sprite.  I've never used weakrefs myself, but this might be a possibility.  I didn't see an implementation of a weakref list (for `.animations`) in the docs in my glance, but I have to imagine it is possible.

And, thanks for the reminder on `kill()`.  My example was just thrown together.  But, yes, the `kill()` function is the way to delete sprites.

**Comment by maartendp** [comment]
A weakref is also a good way to go, and I have actually used this solution in a similar project I did.

The problem I'm having with weakrefs is that it's very implicit.
It can be hard to keep track of sprites and the scopes they're living in, especially the bigger your game gets and the more intricate your logic becomes. It's very likely that something somewhere is keeping an unwanted ref to your sprite, and a weakref won't be a guaranteed solution to the the problem you're having above.

Hence I would prefer making it explicit by having people use the `kill` function.

That being said, though: there is no reason **not** to use weakrefs :) 

I think I will approach this in two steps:
- Implement the kill logic in a first instance
- Migrate to using weakrefs for the event manager and the animation manager, at a later stage.

The reason I think weakrefs are still a good way to go is because I feel my library should be non-intrusive, and keeping handles on refs is all but non-intrusive.

 I will still advocate for using the kill function in the readme, though.

Thank you for all the great ideas. I hope I'll have some time this weekend to look into it.

**Comment by SirGnip** [comment]
Solid reasoning. 

You bring up the point about object lifetimes/scopes getting complex as games scale up.  Agreed.  
To your point about "very likely that something somewhere is keeping an unwanted ref to your sprite, and a weakref won't be a guaranteed solution to the the problem you're having above."  Also agreed.  Weakref won't prevent people from making mistakes.  But, I think curtains (a scene/event/animation library) shouldn't bend over backwards to guess what the right thing is to do if a game is being sloppy with object lifetime management.  If curtains uses weakrefs, you assume that an animation object should live as long as the object it is attached to, period.  If a dev "meant" for an object to be destroyed but still has a reference to it... and curtains is holding extra animation objects around because of it... that isn't a problem in curtains... that is a bug in the game code.

That being said, I think starting with a `kill()` implementation and moving to a weakref implementation potentially later is a good approach.  `kill()` is very explicit (even though it can be forgotten) and easier to understand.

(And, for the record, I think the "non-intrusive" nature of curtains is great!  I started using curtains because of the fire-and-forget animation!)

**Comment by maartendp** [comment]
Hey SirGnip,

Can you install 0.2.1 and give it a try?
Let me know if it's what you expected.

regards,
Maarten

**Comment by SirGnip** [comment]
Sure, I'll try it later and let you know.

**Comment by SirGnip** [comment]
I gave 0.2.1 a try and it worked!  Thanks so much for the fix!

You can see the fix here on a simple project of mine that I'm tinkering with: https://github.com/SirGnip/swiveling_arcs_screensaver/commit/2b273380baf5c5095044fc2e39a5f3647dd10ee6

Its worth mentioning that I'm not using curtain's animation features exactly how they are intended (my apologies 😏 ).  I'm relying heavily on duck typing...  I had something I wanted to animate, but it wasn't an `arcade.Sprite`.  So, I made it work by assigning my own `AnimationManagerProxy` to my object, making sure the properties I'm animating have names found in`TRACKED_ATTRIBUTES`, etc.  And, now I'm explicitly calling `animate.kill()` since I can't rely on `Sprite.kill()`.  But, it works, I get your animation leak fix, and I'm not too deep into customization of curtains.

I was thinking of adding a feature request for curtains to suggest generalization of the library to support animation of any object (not just `arcade.Sprite`).  But, not a big deal.  I'm just tinkering around right now...

So, thanks for the fix!

(Curious to see how the linear search for killing animations scales with regard to performance... But, not a concern for no

```
def kill(self, sprite):
    for animator in self.animations[:]:
            if animator.kill(sprite):```

**Comment by maarten-dp** [comment]
I must admit, I've been thinking about making curtains library agnostic (to a certain degree, at some point you just have to add some default bindings).
You're more than welcome to open this feature request, and perhaps add the pain points you've encountered so far.

As for the linear killing, I was thinking about making a registry, but felt like it added too much complexity. This fix was very straightforward, and I assumed (perhaps wrongly) that you wouldn't be killing sprites each frame.

If you feel like the performance is subpar, feel free to let me know.

**Comment by maarten-dp** [comment]
fixed in 0.2.1
